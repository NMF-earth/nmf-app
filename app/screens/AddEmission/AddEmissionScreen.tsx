import React, { useState } from "react";
import { View, ScrollView } from "react-native";
import { TransportEnum, FoodEnum } from "carbon-footprint";
import { Text, Tag } from "../../components";
import styles from "./AddEmissionScreen.styles";
import navigationOptions from "./AddEmissionScreen.navigationOptions";
import { t } from "../../utils";
import { Food, Transport, Custom, AddEmissionButton } from "./components";
import { EmissionEnum } from "../../interfaces";
import { EmissionPayload } from "../../interfaces/emission/emission.interface";

interface Props {
  navigation: {
    push: (screen: string) => void;
    goBack: () => void;
  };
}

const AddEmissionScreen: React.FunctionComponent<Props> & {
  navigationOptions: typeof navigationOptions;
} = ({ navigation }) => {
  const [emissionType, setEmissionType] = useState(EmissionEnum.transport);
  const [transportType, setTransportType] = useState(TransportEnum.car);
  const [foodType, setFoodType] = useState(FoodEnum.redMeat);
  const [co2eqKilograms, setCo2eqKilograms] = useState(0);
  const [durationHours, setDurationHours] = useState(0.5);
  const [distance, setDistance] = useState(50);
  const [quantity, setQuantity] = useState(0);

  const emissionPayload: EmissionPayload = {
    emissionType: emissionType,
    value: 0,
    emissionModelType: null
  };

  const renderTransport = () => {
    if (emissionType === EmissionEnum.transport) {
      if (transportType === TransportEnum.plane) {
        emissionPayload.value = durationHours;
        emissionPayload.emissionModelType = transportType;
      } else {
        emissionPayload.value = distance;
        emissionPayload.emissionModelType = transportType;
      }
      return (
        <Transport
          setDistanceKilometers={setDistance}
          setDurationHours={setDurationHours}
          setCo2eqKilograms={setCo2eqKilograms}
          setTransportType={setTransportType}
          transportType={transportType}
        />
      );
    }
    return null;
  };

  const renderFood = () => {
    if (emissionType === EmissionEnum.food) {
      emissionPayload.value = quantity;
      emissionPayload.emissionModelType = foodType;
      return (
        <Food
          setQuantityKilograms={setQuantity}
          setCo2eqKilograms={setCo2eqKilograms}
          setFoodType={setFoodType}
          foodType={foodType}
        />
      );
    }
    return null;
  };

  const renderCustom = () => {
    if (emissionType === EmissionEnum.custom) {
      emissionPayload.value = co2eqKilograms;
      emissionPayload.emissionModelType = "custom";
      return <Custom setCo2eqKilograms={setCo2eqKilograms} />;
    }
    return null;
  };

  return (
    <ScrollView style={styles.container}>
      <View style={styles.typeContainer}>
        <View style={styles.textContainer}>
          <Text.H3>{t("ADD_EMISSION_EMISSION_TYPE")}</Text.H3>
        </View>
        <ScrollView horizontal style={styles.tagContainer}>
          <Tag
            icon={"md-airplane"}
            selected={emissionType === EmissionEnum.transport}
            title={"Transport"}
            onPress={() => setEmissionType(EmissionEnum.transport)}
          />
          <Tag
            icon={"md-restaurant"}
            selected={emissionType === EmissionEnum.food}
            title={"Food"}
            onPress={() => setEmissionType(EmissionEnum.food)}
          />
          <Tag
            icon={"md-build"}
            selected={emissionType === EmissionEnum.custom}
            title={"Custom"}
            onPress={() => setEmissionType(EmissionEnum.custom)}
          />
        </ScrollView>
      </View>

      {renderTransport()}
      {renderFood()}
      {renderCustom()}

      <AddEmissionButton
        navigation={navigation}
        emissionPayload={emissionPayload}
      />
    </ScrollView>
  );
};

AddEmissionScreen.navigationOptions = navigationOptions;

export default AddEmissionScreen;
