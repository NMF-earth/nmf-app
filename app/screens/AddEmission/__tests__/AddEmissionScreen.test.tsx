import React from "react";
import { create } from "react-test-renderer";
import { GlobalizeProvider } from "react-native-globalize";

import AddEmissionScreen from "../AddEmissionScreen";

const RealDate = Date.now;

beforeAll(() => {
  Date.now = jest.fn(() => new Date("2019-04-07T10:20:30Z").getTime());
});

afterAll(() => {
  Date.now = RealDate;
});

it("AddEmissionScreen renders correctly", () => {
  const tree = create(
    <GlobalizeProvider locale="en">
      <AddEmissionScreen />
    </GlobalizeProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
