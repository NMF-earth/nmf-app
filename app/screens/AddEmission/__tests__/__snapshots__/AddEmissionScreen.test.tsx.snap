// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddEmissionScreen renders correctly 1`] = `
<GlobalizeProvider
  locale="en"
>
  <KeyboardAwareScrollView
    style={
      Object {
        "alignContent": "flex-start",
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 12,
        }
      }
    >
      <Text.H2
        style={
          Object {
            "marginTop": 4,
            "textTransform": "capitalize",
          }
        }
      />
      <Text.Primary
        lightGray={true}
        style={
          Object {
            "marginTop": 4,
            "textTransform": "capitalize",
          }
        }
      >
        Custom
      </Text.Primary>
    </View>
    <TextInput
      isOptional={true}
      onChangeText={[Function]}
      placeholder="e.g burger, flight from Paris to Romeâ€¦"
      title="Name this emission "
      value=""
    />
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      transparent={true}
      visible={false}
    />
    <View
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 12,
        }
      }
    >
      <Text.H3>
        Date
      </Text.H3>
      <View
        style={
          Object {
            "flexDirection": "column",
            "paddingBottom": 4,
            "paddingTop": 10,
          }
        }
      >
        <TextButton
          iconLeft="calendar"
          onPress={[Function]}
          text="Sunday 7th April 2019"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 12,
        }
      }
    >
      <Text.H3>
        Periodicity
      </Text.H3>
      <View
        style={
          Object {
            "flexDirection": "column",
            "paddingBottom": 4,
            "paddingTop": 10,
          }
        }
      >
        <TextButton
          iconLeft="repeat"
          onPress={[Function]}
          text="Non-recurring"
        />
      </View>
    </View>
    <Button.Primary
      onPress={[Function]}
      style={
        Object {
          "margin": 16,
        }
      }
      text="Add this emission"
    />
  </KeyboardAwareScrollView>
</GlobalizeProvider>
`;
